### PR TITLE
fix(sampling): preserve softmax normalization at low temperatures

### PR DESCRIPTION
**Pull request by Stelath**
## 📌 Description

Fix incorrect softmax normalization at low temperatures. On GB300 with `temperature=0.001`, a row containing one logit of `42` and all others `-20` can produce a probability of `1.0013437` instead of `1.0`. In speculative decoding, this can incorrectly reject a valid token and leave an empty rejection distribution.

Use `__fmul_rn` at all four temperature-scaling sites to round the multiplication consistently across softmax passes, rather than allowing it to fuse with a subsequent subtraction.

Add regression coverage for scalar and per-row temperatures across fused and split-vocabulary execution paths.

## 🔍 Related Issues

No linked issue. This reproduces serially and is separate from the multi-stream workspace race in #4692.

## 🚀 Pull Request Checklist

Thank you for contributing to FlashInfer! Before we review your pull request, please make sure the following items are complete.

### ✅ Pre-commit Checks

- [x] I have installed `pre-commit` by running `pip install pre-commit` (or used your preferred method).
- [x] I have installed the hooks with `pre-commit install`.
- [ ] I have run the hooks manually with `pre-commit run --all-files` and fixed any reported issues.

> If you are unsure about how to set up `pre-commit`, see [the pre-commit documentation](https://pre-commit.com/).

Changed-file hooks passed:
```bash
pre-commit run --files include/flashinfer/sampling.cuh tests/utils/test_sampling.py
```
Repository-wide hooks have not been run locally.

## 🧪 Tests

- [x] Tests have been added or updated as needed.
- [ ] All tests are passing (`unittest`, etc.).

Validated on GB300 (SM103):
- All six new regression cases passed against the corrected installed GPU module, covering three shapes with scalar and per-row temperatures.
- The original split-vocabulary kernel fails normalization; the corrected kernel passes.
- Four captured speculative-decoding failures replay correctly with the corrected probabilities.
- Additional checks passed at temperatures `1e-5`, `0.001`, `0.01`, `1.0`, and `2.0`.

The full upstream suite and cross-architecture validation have not been completed.

**Minimal reproduction**
```python
import torch
import flashinfer

logits = torch.full((5, 129280), -20.0, device="cuda")
logits[:, 123] = 42.0
probs = flashinfer.sampling.softmax(logits, temperature=0.001)
# Before: probs[:, 123] == 1.0013437
# After:  probs[:, 123] == 1.0
```

**Performance:** Maximum measured softmax-kernel overhead was **0.23% (under 0.05 µs)** on GB300 across batch sizes 5–276 at temperatures `0.001` and `1.0`, using FP32 logits, vocabulary size 129280, and CUDA graphs (median of three alternating before/after runs).

## 🔬 Experimental Track

- [ ] This PR is **experimental**: it adds or changes code under `flashinfer/experimental/` and/or an `@flashinfer_experimental_api`. Tracking issue: #
  - [ ] The tracking issue names an owner, the reason for the experimental path, and a graduation plan with a target release.
  - [ ] Core changes are limited to a thin entry point (signature, shared validation, feature-gate check, backend selection, handoff).
  - [ ] Tests live in `tests/experimental/` and were validated on the intended hardware; a runnable example is included.
  - [ ] Nothing is registered in `flashinfer/aot.py`, and no experimental backend is reachable from `backend="auto"` without `FLASHINFER_ALLOW_EXPERIMENTAL_AUTO_BACKENDS=1`. (Calling an `@flashinfer_experimental_api` or naming a backend explicitly is itself the opt-in and needs no environment variable.)
  - [ ] **Test scope declared below.** The experimental CI lane runs exactly these targets, so keep them as narrow as the change allows.

```experimental-tests
# One target per line: a directory or a file. (A pytest ::selector is not
# supported -- the sharding runner cannot consume one.) Must be under
# tests/experimental/ and must exist. Delete these comment lines and add yours, e.g.
#
#   tests/experimental/test_my_backend.py
#   tests/experimental/my_backend/
#
# Declaring the whole tree (tests/experimental/) is allowed but means every
# experimental PR pays for every other feature's tests, in every matrix cell.
```

## Reviewer Notes

When testing header changes, rebuild the sampling module: an installed prebuilt JIT cache can otherwise load the old binary.

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved softmax numerical stability at very low temperatures, preventing probability mass from being incorrectly lost and ensuring dominant logits produce near one-hot results.
  * Applied consistently across supported online softmax execution paths.

* **Tests**
  * Added coverage for low-temperature softmax normalization with both scalar and per-row temperature inputs.

### CODE REVIEW

File: include/flashinfer/sampling.cuh [hunk 1]
```diff
@@ -339,7 +339,8 @@ __global__ FLASHINFER_SAMPLING_LAUNCH_BOUNDS(BLOCK_THREADS) void OnlineSoftmaxFu
 
 #pragma unroll
       for (uint32_t j = 0; j < VEC_SIZE; ++j) {
-        logits_vec[j] *= inv_temp;
+        // Keep max/output scaling identical: do not contract multiplication into FMA.
```
Comment (Stelath) [inline include/flashinfer/sampling.cuh]:
```suggestion

```

File: tests/utils/test_sampling.py [hunk 2]
```diff
@@ -38,6 +38,20 @@ def gumbel_noise(shape, device):
     return gumbel_noise
 
 
+@pytest.mark.parametrize("shape", [(1, 4096), (5, 129280), (230, 129280)])
+@pytest.mark.parametrize("per_row_temperature", [False, True])
+def test_softmax_low_temperature_normalization(shape, per_row_temperature):
+    logits = torch.full(shape, -20.0, device="cuda:0")
+    logits[:, 123] = 42.0
+    temperature = 0.001
```
[the remark this answers is not in the document]
Comment (Stelath) [reply tests/utils/test_sampling.py:46]: The test does detect the bug, this shouldn't be necessary

**Comment by bkryu** [comment]
@flashinfer-bot run

File: include/flashinfer/sampling.cuh
Comment (bkryu) [inline include/flashinfer/sampling.cuh:1]: In this file can you add a comment explaining why.

The diff adds four `__fmul_rn` calls with no comment. To a future reader, `__fmul_rn(x, inv_temp)` might seem overly complicated. A comment like
```
  // __fmul_rn, not *: it must not be contracted into an FMA.
  // *= inv_temp can lead to probabilities exceeding 1 at low temperatures (see PR #5088).
```

**Review by bkryu — COMMENTED** [review]
Thanks @Stelath, no issues with the PR itself, but I think we need to note why we are making this change in the comments.

Left a comment about the comment. Can you help address?

File: include/flashinfer/sampling.cuh
Comment (Stelath) [reply include/flashinfer/sampling.cuh:1]: Done

**Comment by bkryu** [comment]
@flashinfer-bot run

**Review by bkryu — APPROVED** [review]
